src/sqlfluff/core/linter/fix.py

```diff
@@ -0,0 +1,308 @@
+        # Definitely bad if > 2.
+        return False  # pragma: no cover
+
+
+def compute_anchor_edit_info(fixes: List["LintFix"]) -> Dict[int, AnchorEditInfo]:
+    """Group and count fixes by anchor, return dictionary."""
+    anchor_info = defaultdict(AnchorEditInfo)  # type: ignore
+    for fix in fixes:
+        # :TRICKY: Use segment uuid as the dictionary key since
+        # different segments may compare as equal.
+        anchor_id = fix.anchor.uuid
+        anchor_info[anchor_id].add(fix)
+    return dict(anchor_info)
+
+
+def apply_fixes(
+    segment: BaseSegment,
+    dialect: "Dialect",
+    rule_code: str,
+    fixes: Dict[int, AnchorEditInfo],
+) -> Tuple["BaseSegment", List["BaseSegment"], List["BaseSegment"], bool]:
+    """Apply a dictionary of fixes to this segment.
+
+    Used in applying fixes if we're fixing linting errors.
+    If anything changes, this should return a new version of the segment
+    rather than mutating the original.
+
+    Note: We need to have fixes to apply AND this must have children. In the case
+    of raw segments, they will be replaced or removed by their parent and
+    so this function should just return self.
+    """
+    if not fixes or segment.is_raw():
+        return segment, [], [], True
+
+    seg_buffer = []
+    before = []
+    after = []
+    fixes_applied: List[LintFix] = []
+    requires_validate = False
+
+    for seg in segment.segments:
+        # Look for uuid match.
+        # This handles potential positioning ambiguity.
+        anchor_info: Optional[AnchorEditInfo] = fixes.pop(seg.uuid, None)
+
+        if anchor_info is None:
+            # No fix matches here, just add the segment and move on.
+            seg_buffer.append(seg)
+            continue
+
+        # Otherwise there is a fix match.
+        seg_fixes = anchor_info.fixes
+        if (
+            len(seg_fixes) == 2 and seg_fixes[0].edit_type == "create_after"
+        ):  # pragma: no cover
+            # Must be create_before & create_after. Swap so the
+            # "before" comes first.
+            seg_fixes.reverse()
+
+        for f in anchor_info.fixes:
+            assert f.anchor.uuid == seg.uuid
+            fixes_applied.append(f)
+            linter_logger.debug(
+                "Matched fix for %s against segment: %s -> %s",
+                rule_code,
+                f,
+                seg,
+            )
+
+            # Deletes are easy.
+            if f.edit_type == "delete":
+                # We're just getting rid of this segment.
+                requires_validate = True
+                # NOTE: We don't add the segment in this case.
+                continue
+
+            # Otherwise it must be a replace or a create.
+            assert f.edit_type in (
+                "replace",
+                "create_before",
+                "create_after",
+            ), f"Unexpected edit_type: {f.edit_type!r} in {f!r}"
```

Comment:
Wherever it's created, could be created as union of literals instead of a plain string, which would avoid a runtime check?

Reply:
This is a _GREAT_ idea. Will see if I can get that to work.

Reply:
Hmmmm.... actually at this point in the code it's kind of useful as documentation. It's also not on the critical path so I'll keep it as is for now. Neat idea though 👍